File: 3.longest-substring-without-repeating-characters.java
/*
 * @lc app=leetcode id=3 lang=java
 *
 * [3] Longest Substring Without Repeating Characters
 */

// @lc code=start

import java.util.HashMap;

class Solution {
    public int lengthOfLongestSubstring(String s) {
        int left = 0, right = 0, max = 0;
        HashMap<Character, Integer> map = new HashMap<Character, Integer>();
       while(right < s.length()) {
        char c = s.charAt(right);
        if(map.containsKey(c) && map.get(c) >= left) {
            left = map.get(c) + 1;
        }
        map.put(c,right);
        max = Math.max(max, right -left + 1);
        right++;
       }
        return max;
    }
}
// @lc code=end

File: js/3.longest-substring-without-repeating-characters.js
/*
 * @lc app=leetcode id=3 lang=javascript
 *
 * [3] Longest Substring Without Repeating Characters
 */

// @lc code=start
/**
 * @param {string} s
 * @return {number}
 */
var lengthOfLongestSubstring = function (s) {
    let start = 0, max = 0;
    let map = new Map();

    for (let end = 0; end < s.length; end++) {
        if (map.has(s[end]) && map.get(s[end]) >= start) {
            start = map.get(s[end]) + 1;
        }
        map.set(s[end], end);
        max = Math.max(max, end - start + 1);
    }
    return max;
};
// @lc code=end

